src/coreclr/jit/valuenum.cpp

```diff
@@ -2134,6 +2134,16 @@ ValueNum ValueNumStore::VNForFunc(var_types typ, VNFunc func, ValueNum arg0VN, V
         return EvalFuncForConstantArgs(typ, func, arg0VN, arg1VN);
     }
 
+    if ((genTreeOps(func) == GT_EQ) && (arg0VN == arg1VN))
+    {
+        return VNForIntCon(1);
+    }
+
+    if ((genTreeOps(func) == GT_NE) && (arg0VN == arg1VN))
+    {
+        return VNForIntCon(0);
+    }
+
```

Comment:
I assume we can do more if it catches diffs e.g. OR(vn0, vn1) -> vn0 for vn0 == vn1 etc 🙂

Reply:
Yep.

Also I need to make sure this does not fold `float.NaN == float.NaN`...